app/scripts/background.js

```diff
@@ -67,8 +67,8 @@ if (inTest || process.env.METAMASK_DEBUG) {
   global.metamaskGetState = localStore.get.bind(localStore);
 }
 
-// initialization flow
-initialize().catch(log.error);
+const initApp = initialize().catch(log.error);
+extension.runtime.onConnect.addListener(initApp);
```

Comment:
Having an event handler at top level is helpful for MV3, I found that at times in MV3 the first event was missed is user opened extension immediately after installing it.

The code above is good for MV2 also, An optimisation we should do here is, initialise the controllers outside the listeners. Reactive programming will be useful here.